datafusion/expr/src/udf.rs

```diff
@@ -79,7 +90,11 @@ impl std::hash::Hash for ScalarUDF {
 }
 
 impl ScalarUDF {
-    /// Create a new ScalarUDF
+    /// Create a new ScalarUDF from low level details.
+    ///
+    /// See  [`ScalarUDFImpl`] for a more convenient way to create a
+    /// `ScalarUDF` using trait objects
+    #[deprecated(since = "34.0.0", note = "please implement ScalarUDFImpl instead")]
```

Comment:
I think this low level API is quite akward to use and very hard to extend in backwards compatible ways. The trait is easer to use and easier to extend. 

Thus I propose marking this API as deprecated (note most of the examples in codebase use `create_udf` rather than `ScalarUDF:new()` directly

Reply:
Agreed that current low-level API looks awkward to use. Ideally a trait defining what a UDF should implement should be better solution.